core/src/main/java/org/infinispan/marshall/core/ClassIdentifiers.java

```diff
@@ -0,0 +1,71 @@
+package org.infinispan.marshall.core;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.container.entries.InternalCacheValue;
+
+public class ClassIdentifiers implements InternalClassIds {
+   private final static ClassIdentifiers INSTANCE = new ClassIdentifiers();
+
+   // the hashmap is not changed after static ctor, therefore concurrent access is safe
+   private final Map<Class<?>, Integer> classToId = new HashMap<>();
```

Comment:
`HashMap` sucks for this. We gave much efficient ways to have Class to Integer mappings, see what was done in `ClassToExternalizerMap`.